e2e/main.sh

```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/env bash
+
+set -o errtrace
+set -o nounset
+set -o pipefail
+
+SCRIPT_DIR="$(readlink -f "$(dirname "${BASH_SOURCE[0]}")")"
+OUTPUT_DIR="${ARTIFACT_DIR:-_output}/microshift-e2e-$(date +'%Y%m%d-%H%M%S')/"
+[ ! -d "${OUTPUT_DIR}" ] && mkdir -p "${OUTPUT_DIR}"
+
+log() {
+    echo -e "$(date +'%H:%M:%S.%N')    $*"
+}
+
+var_should_not_be_empty() {
+    local var_name=${1}
+    if [ -z "${!var_name+x}" ]; then
+        echo >&2 "Environmental variable '${var_name}' is unset"
+        return 1
+    elif [ -z "${!var_name}" ]; then
+        echo >&2 "Environmental variable '${var_name}' is empty"
+        return 1
+    fi
+}
+
+function_should_be_exported() {
+    local fname=${1}
+    if ! declare -F "$fname"; then
+        echo >&2 "Function '$fname' is unexported. It is expected that function is provided for interacting with cloud provider"
+        return 1
+    fi
+}
+
+prechecks() {
+    var_should_not_be_empty USHIFT_IP || exit 1
+    var_should_not_be_empty USHIFT_USER || exit 1
+    # TODO Check passwordless SSH
+    # TODO Check passwordless sudo
+
+    # Just warning for now
+    # Following functions needed only for runs in CI
+    function_should_be_exported firewall::open_port
+    function_should_be_exported firewall::close_port
+}
+
+microshift_get_konfig() {
+    tmpfile=$(mktemp /tmp/microshift-e2e-konfig.XXXXXX)
+    ssh "$USHIFT_USER@$USHIFT_IP" 'sudo cat /var/lib/microshift/resources/kubeadmin/'"$USHIFT_IP"'/kubeconfig' >"$tmpfile"
+    echo "$tmpfile"
+}
+
+microshift_check_readiness() {
+    local test_output="${1}"
+    log "Waiting for MicroShift to become ready"
+    ssh "$USHIFT_USER@$USHIFT_IP" \
+        "sudo /etc/greenboot/check/required.d/40_microshift_running_check.sh | \
```

Comment:
Passing shell commands to ssh to run with pipes and loops and things can make quoting pretty messy. Can we process the output of the script outside of ssh? Or can we make the script being run produce just the output we want (maybe via a command line switch)?

We also, at one point, talked about just copying all of the test code into the VM and then running everything there. I can see some issues with that, but the appeal is that is simplifies the test runner logic, at least in some cases. Does it make the implementation of the tests harder, though?

Reply:
I feel that tests like router/lb smokes would not be as sane ran from within.
Also - reboot test (which I haven't ported(? - there's nothing to port there actually...) yet)

I know that current state of the PR is "in between" approaches, but I didn't wanted to make too many changes to existing tests (esp greenboot) hoping for iterative approach.

Re that monstrosity after pipe - it just adds timestamp to beginning of the line, somehow I couldn't find moreutils package on RHEL (which would provide [ts](https://linux.die.net/man/1/ts) and that would be much simpler).
Also, it's not needed at all - just put it there because I noticed that greenboot check takes way too long (pod restart check takes 50s for each namespace sequentially)

Reply:
I just remembered testing I've done simulating "power loss during initialization" - test similar to this (or just "do 100 reboots")  has no place running from the microshift itself. 

Having external tests runner feels cleaner to me, but there's nothing prohibiting the test suite from the microshift (like selected tests running on dev vm against dev vm), but that's probably not what you meant?
Are we talking about test runner on dev VM or just having test cases in scripts, copied to remote and executed there? Even though bash scripts are much simpler at times due to being native to shell (duh) or `set -x` providing full debug info, they feel like a mess.

Reply:
I'm sorry, yes, you've pointed out that reboot case to me a few times before, and I keep forgetting it.

I'm trying to minimize the complexity of the commands embedded in the ssh arguments. Maybe we can isolate just those cases as temporary scripts? A guideline like "if you wouldn't type it on the command line, don't pass it to ssh as an argument" could give us a heuristic. Maybe that's stricter than needed, though?